File: backend/internal/engine/geometry.go
package engine

import (
	"bytes"
	"encoding/base64"
	"image"
	"image/color"
	"image/png"
	"math"

	"gomiro/internal/model"
)

type AABB struct {
	MinX, MinY, MaxX, MaxY float64
}

func (a AABB) Intersects(b AABB) bool {
	return a.MinX <= b.MaxX && a.MaxX >= b.MinX && a.MinY <= b.MaxY && a.MaxY >= b.MinY
}

func (a AABB) Union(b AABB) AABB {
	return AABB{
		MinX: math.Min(a.MinX, b.MinX),
		MinY: math.Min(a.MinY, b.MinY),
		MaxX: math.Max(a.MaxX, b.MaxX),
		MaxY: math.Max(a.MaxY, b.MaxY),
	}
}

func (a AABB) Empty() bool {
	return a.MaxX < a.MinX || a.MaxY < a.MinY
}

func ShapeAABB(s *model.Shape) AABB {
	if s == nil {
		return AABB{1, 1, -1, -1}
	}
	if len(s.Points) > 0 && (s.Kind == model.KindFreedraw || s.Kind == model.KindLine || s.Kind == model.KindArrow) {
		minX, minY := s.Points[0].X, s.Points[0].Y
		maxX, maxY := minX, minY
		for _, p := range s.Points {
			if p.X < minX {
				minX = p.X
			}
			if p.Y < minY {
				minY = p.Y
			}
			if p.X > maxX {
				maxX = p.X
			}
			if p.Y > maxY {
				maxY = p.Y
			}
		}
		pad := s.StrokeW + 8
		return AABB{minX + s.X - pad, minY + s.Y - pad, maxX + s.X + pad, maxY + s.Y + pad}
	}
	x1, y1 := s.X, s.Y
	x2, y2 := s.X+s.W, s.Y+s.H
	if s.W < 0 {
		x1, x2 = x2, x1
	}
	if s.H < 0 {
		y1, y2 = y2, y1
	}
	if s.Rotation == 0 {
		return AABB{x1, y1, x2, y2}
	}
	cx := (x1 + x2) / 2
	cy := (y1 + y2) / 2
	rad := s.Rotation * math.Pi / 180
	cos, sin := math.Cos(rad), math.Sin(rad)
	corners := [4][2]float64{{x1, y1}, {x2, y1}, {x2, y2}, {x1, y2}}
	minX, minY := math.Inf(1), math.Inf(1)
	maxX, maxY := math.Inf(-1), math.Inf(-1)
	for _, c := range corners {
		dx, dy := c[0]-cx, c[1]-cy
		rx := cx + dx*cos - dy*sin
		ry := cy + dx*sin + dy*cos
		if rx < minX {
			minX = rx
		}
		if ry < minY {
			minY = ry
		}
		if rx > maxX {
			maxX = rx
		}
		if ry > maxY {
			maxY = ry
		}
	}
	return AABB{minX, minY, maxX, maxY}
}

func ContentBounds(shapes map[string]*model.Shape) AABB {
	var acc AABB
	first := true
	for _, s := range shapes {
		if s == nil || s.Deleted {
			continue
		}
		b := ShapeAABB(s)
		if first {
			acc = b
			first = false
			continue
		}
		acc = acc.Union(b)
	}
	if first {
		return AABB{-400, -300, 400, 300}
	}
	return acc
}

func AnchorPoint(s *model.Shape, name string) (float64, float64) {
	if s == nil {
		return 0, 0
	}
	cx, cy := s.X+s.W/2, s.Y+s.H/2
	switch name {
	case "n":
		return cx, s.Y
	case "s":
		return cx, s.Y + s.H
	case "w":
		return s.X, cy
	case "e":
		return s.X + s.W, cy
	case "nw":
		return s.X, s.Y
	case "ne":
		return s.X + s.W, s.Y
	case "sw":
		return s.X, s.Y + s.H
	case "se":
		return s.X + s.W, s.Y + s.H
	default:
		return cx, cy
	}
}

func Distance(ax, ay, bx, by float64) float64 {
	dx, dy := ax-bx, ay-by
	return math.Hypot(dx, dy)
}

func PointInAABB(x, y float64, b AABB) bool {
	return x >= b.MinX && x <= b.MaxX && y >= b.MinY && y <= b.MaxY
}

func PointInRotatedRect(x, y float64, s *model.Shape) bool {
	if s == nil {
		return false
	}
	if s.Rotation == 0 {
		return PointInAABB(x, y, ShapeAABB(s))
	}
	cx, cy := s.X+s.W/2, s.Y+s.H/2
	rad := -s.Rotation * math.Pi / 180
	cos, sin := math.Cos(rad), math.Sin(rad)
	dx, dy := x-cx, y-cy
	lx := dx*cos - dy*sin
	ly := dx*sin + dy*cos
	hw, hh := math.Abs(s.W)/2, math.Abs(s.H)/2
	return math.Abs(lx) <= hw && math.Abs(ly) <= hh
}

func PointInEllipse(x, y float64, s *model.Shape) bool {
	if s == nil || s.W == 0 || s.H == 0 {
		return false
	}
	cx, cy := s.X+s.W/2, s.Y+s.H/2
	nx := (x - cx) / (s.W / 2)
	ny := (y - cy) / (s.H / 2)
	return nx*nx+ny*ny <= 1
}

func PointInDiamond(x, y float64, s *model.Shape) bool {
	if s == nil || s.W == 0 || s.H == 0 {
		return false
	}
	cx, cy := s.X+s.W/2, s.Y+s.H/2
	return math.Abs((x-cx)/(s.W/2))+math.Abs((y-cy)/(s.H/2)) <= 1
}

func DistToSegment(px, py, ax, ay, bx, by float64) float64 {
	dx, dy := bx-ax, by-ay
	if dx == 0 && dy == 0 {
		return Distance(px, py, ax, ay)
	}
	t := ((px-ax)*dx + (py-ay)*dy) / (dx*dx + dy*dy)
	if t < 0 {
		t = 0
	} else if t > 1 {
		t = 1
	}
	return Distance(px, py, ax+t*dx, ay+t*dy)
}

func HitTest(s *model.Shape, x, y, pad float64) bool {
	if s == nil || s.Deleted {
		return false
	}
	switch s.Kind {
	case model.KindEllipse:
		return PointInEllipse(x, y, s)
	case model.KindDiamond:
		return PointInDiamond(x, y, s)
	case model.KindLine, model.KindArrow, model.KindFreedraw:
		return hitStroke(s, x, y, pad)
	default:
		return PointInRotatedRect(x, y, s)
	}
}

func hitStroke(s *model.Shape, x, y, pad float64) bool {
	tol := s.StrokeW/2 + pad
	if len(s.Points) >= 2 {
		for i := 1; i < len(s.Points); i++ {
			a, b := s.Points[i-1], s.Points[i]
			if DistToSegment(x, y, a.X+s.X, a.Y+s.Y, b.X+s.X, b.Y+s.Y) <= tol {
				return true
			}
		}
		return false
	}
	return DistToSegment(x, y, s.X, s.Y, s.X+s.W, s.Y+s.H) <= tol
}

func inflate(a AABB, pad float64) AABB {
	return AABB{a.MinX - pad, a.MinY - pad, a.MaxX + pad, a.MaxY + pad}
}

func NearestAnchor(s *model.Shape, x, y float64) string {
	if s == nil {
		return "c"
	}
	names := []string{"n", "s", "w", "e", "nw", "ne", "sw", "se", "c"}
	best, bestD := "c", math.Inf(1)
	for _, n := range names {
		ax, ay := AnchorPoint(s, n)
		d := Distance(x, y, ax, ay)
		if d < bestD {
			bestD = d
			best = n
		}
	}
	return best
}

func PickTop(shapes map[string]*model.Shape, x, y, pad float64) *model.Shape {
	ids := SortedIDs(shapes)
	for i := len(ids) - 1; i >= 0; i-- {
		s := shapes[ids[i]]
		if HitTest(s, x, y, pad) {
			return s
		}
	}
	return nil
}

func MarqueeHits(shapes map[string]*model.Shape, box AABB) []string {
	var out []string
	for _, s := range shapes {
		if s == nil || s.Deleted {
			continue
		}
		if ShapeAABB(s).Intersects(box) {
			out = append(out, s.ID)
		}
	}
	return out
}

const thumbW, thumbH = 320, 200

// ThumbnailPNG draws a schematic of the board into a 320×200 PNG data URL.
func ThumbnailPNG(shapes map[string]*model.Shape) string {
	world := ContentBounds(shapes)
	ww := world.MaxX - world.MinX
	wh := world.MaxY - world.MinY
	if ww < 1 {
		ww = 1
	}
	if wh < 1 {
		wh = 1
	}
	pad := 16.0
	sx := (float64(thumbW) - pad*2) / ww
	sy := (float64(thumbH) - pad*2) / wh
	scale := math.Min(sx, sy)
	img := image.NewRGBA(image.Rect(0, 0, thumbW, thumbH))
	bg := color.RGBA{0x16, 0x14, 0x11, 0xff}
	ink := color.RGBA{0xc4, 0xa3, 0x5a, 0xff}
	for y := 0; y < thumbH; y++ {
		for x := 0; x < thumbW; x++ {
			img.SetRGBA(x, y, bg)
		}
	}
	project := func(x, y float64) (int, int) {
		px := pad + (x-world.MinX)*scale
		py := pad + (y-world.MinY)*scale
		return int(px), int(py)
	}
	ids := SortedIDs(shapes)
	if len(ids) > 400 {
		ids = ids[len(ids)-400:]
	}
	for _, id := range ids {
		s := shapes[id]
		if s == nil {
			continue
		}
		b := ShapeAABB(s)
		x1, y1 := project(b.MinX, b.MinY)
		x2, y2 := project(b.MaxX, b.MaxY)
		strokeRect(img, x1, y1, x2, y2, ink)
	}
	var buf bytes.Buffer
	if err := png.Encode(&buf, img); err != nil {
		return ""
	}
	return "data:image/png;base64," + base64.StdEncoding.EncodeToString(buf.Bytes())
}

func strokeRect(img *image.RGBA, x1, y1, x2, y2 int, c color.RGBA) {
	if x2 < x1 {
		x1, x2 = x2, x1
	}
	if y2 < y1 {
		y1, y2 = y2, y1
	}
	if x1 < 0 {
		x1 = 0
	}
	if y1 < 0 {
		y1 = 0
	}
	if x2 >= thumbW {
		x2 = thumbW - 1
	}
	if y2 >= thumbH {
		y2 = thumbH - 1
	}
	for x := x1; x <= x2; x++ {
		img.Set(x, y1, c)
		img.Set(x, y2, c)
	}
	for y := y1; y <= y2; y++ {
		img.Set(x1, y, c)
		img.Set(x2, y, c)
	}
}

File: frontend-user/src/lib/geometry.ts
import type { Point, Shape } from './protocol'

export type AABB = { minX: number; minY: number; maxX: number; maxY: number }
export type Viewport = { x: number; y: number; scale: number }
export type Guide = { kind: 'v' | 'h'; at: number; from: number; to: number }

export const MIN_SCALE = 0.1
export const MAX_SCALE = 5

export function emptyAABB(): AABB {
  return { minX: 1, minY: 1, maxX: -1, maxY: -1 }
}

export function aabbEmpty(a: AABB): boolean {
  return a.maxX < a.minX || a.maxY < a.minY
}

export function intersects(a: AABB, b: AABB): boolean {
  return a.minX <= b.maxX && a.maxX >= b.minX && a.minY <= b.maxY && a.maxY >= b.minY
}

export function union(a: AABB, b: AABB): AABB {
  return {
    minX: Math.min(a.minX, b.minX),
    minY: Math.min(a.minY, b.minY),
    maxX: Math.max(a.maxX, b.maxX),
    maxY: Math.max(a.maxY, b.maxY),
  }
}

export function inflate(a: AABB, pad: number): AABB {
  return { minX: a.minX - pad, minY: a.minY - pad, maxX: a.maxX + pad, maxY: a.maxY + pad }
}

export function pointInAABB(x: number, y: number, b: AABB): boolean {
  return x >= b.minX && x <= b.maxX && y >= b.minY && y <= b.maxY
}

export function worldToScreen(vx: number, vy: number, vp: Viewport): { x: number; y: number } {
  return { x: (vx - vp.x) * vp.scale, y: (vy - vp.y) * vp.scale }
}

export function screenToWorld(sx: number, sy: number, vp: Viewport): { x: number; y: number } {
  return { x: sx / vp.scale + vp.x, y: sy / vp.scale + vp.y }
}

export function zoomAt(vp: Viewport, sx: number, sy: number, next: number): Viewport {
  const scale = Math.min(MAX_SCALE, Math.max(MIN_SCALE, next))
  const w = screenToWorld(sx, sy, vp)
  return { x: w.x - sx / scale, y: w.y - sy / scale, scale }
}

export function viewportAABB(vp: Viewport, w: number, h: number): AABB {
  const a = screenToWorld(0, 0, vp)
  const b = screenToWorld(w, h, vp)
  return { minX: a.x, minY: a.y, maxX: b.x, maxY: b.y }
}

export function shapeAABB(s: Shape): AABB {
  if (s.points && s.points.length && (s.kind === 'freedraw' || s.kind === 'line' || s.kind === 'arrow')) {
    let minX = s.points[0].x
    let minY = s.points[0].y
    let maxX = minX
    let maxY = minY
    for (const p of s.points) {
      minX = Math.min(minX, p.x)
      minY = Math.min(minY, p.y)
      maxX = Math.max(maxX, p.x)
      maxY = Math.max(maxY, p.y)
    }
    const pad = (s.strokeW || 0) + 8
    return { minX: minX + s.x - pad, minY: minY + s.y - pad, maxX: maxX + s.x + pad, maxY: maxY + s.y + pad }
  }
  let x1 = s.x
  let y1 = s.y
  let x2 = s.x + s.w
  let y2 = s.y + s.h
  if (s.w < 0) [x1, x2] = [x2, x1]
  if (s.h < 0) [y1, y2] = [y2, y1]
  if (!s.rotation) return { minX: x1, minY: y1, maxX: x2, maxY: y2 }
  const cx = (x1 + x2) / 2
  const cy = (y1 + y2) / 2
  const rad = (s.rotation * Math.PI) / 180
  const cos = Math.cos(rad)
  const sin = Math.sin(rad)
  const corners = [
    [x1, y1],
    [x2, y1],
    [x2, y2],
    [x1, y2],
  ]
  let minX = Infinity
  let minY = Infinity
  let maxX = -Infinity
  let maxY = -Infinity
  for (const [x, y] of corners) {
    const dx = x - cx
    const dy = y - cy
    const rx = cx + dx * cos - dy * sin
    const ry = cy + dx * sin + dy * cos
    minX = Math.min(minX, rx)
    minY = Math.min(minY, ry)
    maxX = Math.max(maxX, rx)
    maxY = Math.max(maxY, ry)
  }
  return { minX, minY, maxX, maxY }
}

export function contentBounds(shapes: Iterable<Shape>): AABB {
  let acc: AABB | null = null
  for (const s of shapes) {
    if (s.deleted) continue
    const b = shapeAABB(s)
    acc = acc ? union(acc, b) : b
  }
  return acc ?? { minX: -400, minY: -300, maxX: 400, maxY: 300 }
}

export function anchorPoint(s: Shape, name: string): { x: number; y: number } {
  const cx = s.x + s.w / 2
  const cy = s.y + s.h / 2
  switch (name) {
    case 'n':
      return { x: cx, y: s.y }
    case 's':
      return { x: cx, y: s.y + s.h }
    case 'w':
      return { x: s.x, y: cy }
    case 'e':
      return { x: s.x + s.w, y: cy }
    case 'nw':
      return { x: s.x, y: s.y }
    case 'ne':
      return { x: s.x + s.w, y: s.y }
    case 'sw':
      return { x: s.x, y: s.y + s.h }
    case 'se':
      return { x: s.x + s.w, y: s.y + s.h }
    default:
      return { x: cx, y: cy }
  }
}

export const ANCHORS = ['n', 's', 'w', 'e', 'nw', 'ne', 'sw', 'se', 'c'] as const

export function nearestAnchor(s: Shape, x: number, y: number): string {
  let best = 'c'
  let bestD = Infinity
  for (const n of ANCHORS) {
    const p = anchorPoint(s, n)
    const d = Math.hypot(x - p.x, y - p.y)
    if (d < bestD) {
      bestD = d
      best = n
    }
  }
  return best
}

export function snapThreshold(scale: number): number {
  if (scale <= 0) return 8
  return Math.min(24, Math.max(2, 8 / scale))
}

export function snapMove(moving: AABB, others: AABB[], threshold: number): { x: number; y: number; guides: Guide[]; snapped: boolean } {
  const th = threshold > 0 ? threshold : 8
  const res = { x: 0, y: 0, guides: [] as Guide[], snapped: false }
  let bestX = th + 1
  let bestY = th + 1
  const mx = [moving.minX, (moving.minX + moving.maxX) / 2, moving.maxX]
  const my = [moving.minY, (moving.minY + moving.maxY) / 2, moving.maxY]
  for (const o of others) {
    const ox = [o.minX, (o.minX + o.maxX) / 2, o.maxX]
    const oy = [o.minY, (o.minY + o.maxY) / 2, o.maxY]
    for (const a of mx) {
      for (const b of ox) {
        const d = b - a
        const ad = Math.abs(d)
        if (ad < bestX && ad <= th) {
          bestX = ad
          res.x = d
          res.guides = res.guides.filter((g) => g.kind !== 'v')
          res.guides.push({ kind: 'v', at: b, from: Math.min(moving.minY, o.minY), to: Math.max(moving.maxY, o.maxY) })
        }
      }
    }
    for (const a of my) {
      for (const b of oy) {
        const d = b - a
        const ad = Math.abs(d)
        if (ad < bestY && ad <= th) {
          bestY = ad
          res.y = d
          res.guides = res.guides.filter((g) => g.kind !== 'h')
          res.guides.push({ kind: 'h', at: b, from: Math.min(moving.minX, o.minX), to: Math.max(moving.maxX, o.maxX) })
        }
      }
    }
  }
  res.snapped = bestX <= th || bestY <= th
  if (bestX > th) res.x = 0
  if (bestY > th) res.y = 0
  return res
}

export function rotateAround(x: number, y: number, cx: number, cy: number, deg: number): { x: number; y: number } {
  const rad = (deg * Math.PI) / 180
  const cos = Math.cos(rad)
  const sin = Math.sin(rad)
  const dx = x - cx
  const dy = y - cy
  return { x: cx + dx * cos - dy * sin, y: cy + dx * sin + dy * cos }
}

export function toLocal(x: number, y: number, s: Shape): { x: number; y: number } {
  const cx = s.x + s.w / 2
  const cy = s.y + s.h / 2
  const p = rotateAround(x, y, cx, cy, -(s.rotation || 0))
  return { x: p.x - s.x, y: p.y - s.y }
}

export function nextZ(shapes: Iterable<Shape>): number {
  let max = 0
  for (const s of shapes) {
    if (!s.deleted && s.z > max) max = s.z
  }
  return max + 1
}

export function sortedShapes(shapes: Iterable<Shape>): Shape[] {
  return [...shapes].filter((s) => !s.deleted).sort((a, b) => (a.z === b.z ? a.id.localeCompare(b.id) : a.z - b.z))
}

export function gridSpec(scale: number): { step: number; mode: 'dot' | 'line' } {
  if (scale >= 1.15) return { step: 20, mode: 'line' }
  if (scale >= 0.55) return { step: 40, mode: 'line' }
  if (scale >= 0.28) return { step: 80, mode: 'dot' }
  return { step: 160, mode: 'dot' }
}

export function worldPoints(s: Shape): Point[] {
  if (s.points && s.points.length) {
    return s.points.map((p) => ({ x: p.x + s.x, y: p.y + s.y, p: p.p }))
  }
  return [
    { x: s.x, y: s.y },
    { x: s.x + s.w, y: s.y + s.h },
  ]
}

export function clamp(n: number, lo: number, hi: number): number {
  return Math.min(hi, Math.max(lo, n))
}
